src/lib/rule2F_test.ts

```diff
@@ -0,0 +1,215 @@
+              </div>
+            </div>
+          </div>
+          <div>!</div>
+        </label>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Hello world !');
+  });
+
+  it('considers CSS generated text content', () => {
+    render(
+      html`
+        <div id="foo">
+          world
+          <div>
+            <style>
+              #foo:before {
+                content: 'Hello';
+              }
+              #foo:after {
+                content: '!';
+              }
+            </style>
+          </div>
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Helloworld!');
+  });
+
+  it('considers entire subtree for descendants of label reference', () => {
+    render(
+      html`
+        <div id="foo">
+          Hello
+          <div aria-labelledby="bar"></div>
+        </div>
+        <div id="bar">
+          world
+          <div aria-labelledby="foo"></div>
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Hello world');
+  });
+});
+
+describe('The function allowsNameFromContent', () => {
+  let container: HTMLElement;
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  it('returns true for roles that allow name from content', () => {
+    render(
+      html`
+        <div id="foo" role="button">
+          Hello world
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('returns false for roles that do not allow name from content', () => {
+    render(
+      html`
+        <div id="foo" role="presentation">
+          Hello world
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(false);
+  });
+
+  it('recognises semantic html elements', () => {
+    render(
+      html`
+        <h1 id="foo">
+          Hello world
+        </h1>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows table calls only if they are within a table', () => {
+    render(
+      html`
+        <table>
+          <td id="foo"></td>
+        </table>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows options only if they are within a datalist or select', () => {
```

Comment:
Try to make the string descriptons exactly match what is being tested. When these tests fail the failure message will be something like 
`Failed: The function allowsNameFromContent allows options only if they are within a datalist or select`.

Which works, but something like this would immediately tell you what the problem is:
`Failed: The function allowsNameFromContent returns false for options if they're not within a datalist or select`.

This aplies throughout this file